File: 2024/11/main.py
import aiofiles
import asyncio
import numpy as np
import time
import tqdm
import functools

def count_stones(stone, cache, n=75):
    if ((stone, n, ) in cache.keys()):
        return cache[(stone, n)]
    
    if n == 0:
        return 1 
    
    # Rule 1
    if stone == 0:
        value = count_stones(1, cache, n - 1)
        cache[(stone, n)] = value
        return value

    # Rule 2
    if len(str(stone)) % 2 == 0:
        stone = str(stone)
        v1 = count_stones(int(stone[len(stone)//2:]), cache, n - 1)
        v2 = count_stones(int(stone[:len(stone)//2]), cache, n - 1)
        value = v1 + v2
        cache[(stone, n)] = value
        return value

    # Rule 3
    value = count_stones(stone * 2024, cache, n - 1)
    cache[(stone, n)] = value
    return value

async def main():
    async with aiofiles.open('input.txt', 'r') as f:
        async for line in f:
            stones = list(map(int, line.strip().split(' ')))

    print(stones)
    cache = dict()
    print(sum([count_stones(s, cache) for s in stones]))
    print(len(cache))


if __name__ == '__main__':
    t1 = time.time()
    asyncio.run(main())
    t2 = time.time()
    print(f'Execution time: {(t2 - t1) * 1000 : .2f}ms')

File: 2024/11/main.r
library(tidyverse)

t1 <- Sys.time()

# use enviroment because it has O(1) access time
cache_env <- new.env()

count_rocks <- function(rock, n) {
    # base case
    if (n == 0) {
        return(1)
    }
    
    # cache
    key <- paste(rock, n, sep = "_")
    
    # cache hit
    if (exists(key, envir = cache_env)) {
        return(get(key, envir = cache_env))
    }
    
    # rule 1
    if (rock == 0) {
        result <- count_rocks(1, n - 1)
        assign(key, result, envir = cache_env)
        return(result)
    }
    
    # rule 2
    num_digits <- floor(log10(rock)) + 1
    if (num_digits %% 2 == 0) {
        d <- 10 ^ (num_digits / 2)
        a <- count_rocks(floor(rock / d), n - 1)
        b <- count_rocks(rock %% d, n - 1)
        result <- a + b
        assign(key, result, envir = cache_env)
        return(result)
    }
    
    # rule 3
    result <- count_rocks(rock * 2024, n - 1)
    assign(key, result, envir = cache_env)
    return(result)
}

file_path <- "input.txt"
numbers_list <- read_lines(file_path) %>%
  str_split(" ") %>%                    # Split the line by spaces
  map(as.integer) %>%                   # Convert to integers
  map(~ map(.x, ~count_rocks(.x, 75)))  # Process each number


# Print results
print(paste("Result:", sum(unlist(numbers_list))))

t2 <- Sys.time()
duration <- (t2 - t1) * 1000
print(sprintf("Execution time: %.2f ms", duration))